python/lib/core/dmod/core/dataset.py

```diff
@@ -487,6 +489,360 @@ def unlink_to_dataset(self, dataset: Dataset) -> bool:
+
+    Abstraction for detecting the  of a single data item.  Here, a data item specifically means either
+    :class:`bytes` object with raw data, a :class:`RepeatableReader` object that can read data multiple times, or a
+    :class:`Path` object pointing to a file (not a directory).
+
+    Base type provides a class method :method:`get_type_for_format` that allows for getting an appropriate subtype for
+    detecting in the contexts of a specific :class:`DataFormat`, assuming one exists.
+
+    A subclass may optionally register with this type if the subclass itself is given a ``format_type`` keyword argument
+    containing a :class:`DataFormat` value.  A ``name`` keyword argument may also optionally be provided; by default,
+    the class's name is used.  E.g.:
+
+    ```
+    class DomainDetectorSubtype(ItemDataDomainDetector, format_type=DataFormat.AORC_CSV):
+    ```
+
+    This will register the subclass as a type that can detect domains for data having that format.  This type provides
+    a class method :method:`get_types_for_format` for accessing the collection of registered subclasses for a format.
+    """
+
+    _all_subclasses: Dict[Type['ItemDataDomainDetector'], str] = dict()
+    """ All known subclasses, keyed to its name identifier (which defaults to the class's name). """
+    _detector_registry: Dict[DataFormat, Dict[str, Type['ItemDataDomainDetector']]] = {f: dict() for f in DataFormat}
+    """ 
+    Registered subclasses in the form of an outer dictionary mapping supported format to inner dictionaries, mapping 
+    subclass name to subclass.
+    """
+    _subclass_formats: Dict[Type['ItemDataDomainDetector'], DataFormat] = dict()
+    """ A collection to reverse the format for a particulr subclass. """
+
+    @classmethod
+    def get_data_format(cls) -> Optional[DataFormat]:
+        """
+        Get the registered data format for this subclass of :class:`ItemDataDomainDetector`, if it is registered.
+
+        Returns
+        -------
+        Optional[DataFormat]
+            The :class:`DataFormat` of this subclass, if it registered one; otherwise ``None``.
+        """
+        return cls._subclass_formats.get(cls)
+
+    @classmethod
+    def get_types_for_format(cls, data_format: DataFormat) -> Dict[str, Type['ItemDataDomainDetector']]:
+        """
+        Return subclass type registered as capable of detecting domains for data with the given data format.
+
+        Parameters
+        ----------
+        data_format: DataFormat
+            The data format of interest.
+
+        Returns
+        -------
+        Dict[str,Type['ItemDataDomainDetector']]
+            Dictionary (keyed by class name) of registered subclasses capable of detecting domains for data with the
+            given data format.
+        """
+        return cls._detector_registry.get(data_format, dict())
+
+    def __init_subclass__(cls, format_type: Optional[DataFormat] = None, name: Optional[str] = None, **kwargs):
```

Comment:
Try to avoid this pattern if at all possible. Consider a wrapper object used to keep track of these (like a sort of manager) rather to mutating shared state upon construction. It will make testing and isolating errors easier.

Reply:
Yeah I second, @christophertubbs here for slightly different reasons. Tracking subtypes and their supported format type at the class level adds complexity and overly couples a type to its behavior. Like @christophertubbs alluded to, due to the nature of python imports, if a subtype is not parsed during runtime, it will not be tracked by the super type. We generally don't have this issue due to our high use of module level exports in `__init__.py`. However, this does not rule out the potential for thorny bugs that are difficult to pin down due, for the sake of example, to the low cohesion between the subtype tracking feature and the implemented subtype. 

While writing procedural style code to register formats to behavior is tedious to write, it is far easier to read and reason with. It is also easier to create feature flags and add new functionality in the future. I recommend that if a format detector factory is necessary, we create that abstraction where it is needed, preferably not in the `core.dataset` module.

Reply:
> Try to avoid this pattern if at all possible.

@christophertubbs, do you mean using `__init_subclass__` or generally tracking subclasses mapped to use case (here, applicable data format) in a superclass?

> Tracking subtypes and their supported format type at the class level adds complexity ... 

@aaraney, can you elaborate further?  Assuming for a moment that we are tracking those details, some amount complexity has to go somewhere.  Is it drastically more by putting it here than in another class/object (which we'd then have to worry explicitly about making available everywhere it needed to be)? 

> ... and overly couples a type to its behavior.

I don't understand.  Are you talking about somehow coupling the superclass to specific subclass behavior?  

> Like @christophertubbs alluded to, due to the nature of python imports, if a subtype is not parsed during runtime, it will not be tracked by the super type. 

This is true, but so is the reverse.  We can take advantage of this to loosen dependency constraints and dynamically utilize new subclasses.  The dependency stuff is my biggest concern, and might require moving things around and introduce more abstract types (so that we don't run into dependency complications with the `dmod.core` package).  Also, the error handling for when something isn't there is a fundamental part of the design, and I'm not sure testing is as difficult as it may first appear.  Though there is the potential for some interesting and subtle problems; I'm not sure how many we'd organically run into.

>  I recommend that if a format detector factory is necessary, we create that abstraction where it is needed, preferably not in the core.dataset module.

Second part first ... sure, I don't have any objections to moving some things into their own module(s).

As for a factory, I think I started that way but ran into a wall:  I don't think there is a feasible way to decide whether a subclass is the subclass we want to create until after we've created it and had it detect a domain (or fail to do so).  

The primary thing we have to represent the general form of some data is the _DataFormat_, so a detector factory would need to use format to pick what subclass of instance to produce.   The format would have to be provided to the factory by whatever code is using the factory.  Thinking a bit higher level for a moment, how will that entity know the data format?  Well, it might be provided before then, but I am pretty sure that implies the user pass it via a CLI arg, and the entire point here is that we don't want to have to rely on that.  

If format won't (always) be provided, then something has to detect it also.  We could do a separate detection, but it's the same data we'd look at to detect the domain.  So why not just detect the entire domain at that point?   For large files especially, we would either fail fast (or at least as fast as possible) or avoid a potentially costly duplicate open and read.

At that point, the only thing a factory would do is encapsulate the process of trying subclasses until we'd found the right one.  But since that must have involved getting an instance of the right subclass to build the thing we actually needed some unknown instance for (which is what we needed a factory for), why not just make the "factory" conform to the interface of the subclass?  And if it is a subclass, why not just keep all the metadata it needs within the superclass, to guarantee access for both the registrants and the "factory?"

(If we do keep this, I might should go back and add something to _UniversalItemDomainDetector_ so that it can save the "right" subclass.  And I probably need to be using the `data_domain` property, rather than `detect`.)

Let me know if any of my reasoning is off or unclear.  We can continue to discuss things here, but it might be worth planning some time to discuss this all together.

Reply:
> @christophertubbs, do you mean using __init_subclass__ or generally tracking subclasses mapped to use case (here, applicable data format) in a superclass?

Yes?

Mostly `__init_subclass__`. It's sort of like using `__new__` - if you're super duper pooper scooper fancy like pandas or Django, sure, have fun with it, but it'll be hard to track, diagnose, and predict. 70% of the time, if you're using `__init_subclass__`, like `__new__`, there's probably an alternate approach to use.

It's hard to track what all the three maps are doing, but that sort of tracking could be done more plainly (i.e. without python magic) through an external record keeper or through a factory. This will help contain volatile data from a global scope (as is being done here). You can also just define an abstract classmethod that will yield that information rather than defining it within the MRO as a kwarg. 

Something I just slapped together:

```python
class Root(abc.ABC):
    _instances = {}
    _format = {}
    _subclasses = {}
    
    @classmethod
    @abc.abstractmethod
    def get_format(cls):
        ...
    def __init__(self, a: str):
        self.a = a
        this_class = self.__class__
        if this_class not in self._instances:
            self._instances[this_class] = 0
        self._instances[this_class] += 1
        self._format[this_class.__name__] = self.get_format()
        self._subclasses[self.get_format()] = self

class Level1(Root):
    @classmethod
    def get_format(cls):
        return "jpg"
    def __init__(self, a: str, b: int):
        super().__init__(a)
        self.b = b
        
class Level2(Root):
    @classmethod
    def get_format(cls):
        return "webp"
    def __init__(self):
        super().__init__(8)
        
l1 = Level1(3, 4)
l2 = Level2()
Level1._instances
# {<class '__main__.Level1'>: 1, <class '__main__.Level2'>: 1}
Level1._format
# {'Level1': 'jpg', 'Level2': 'webp'}
Level2._format
# {'Level1': 'jpg', 'Level2': 'webp'}
Root._format
# {'Level1': 'jpg', 'Level2': 'webp'}
Root._subclasses
# {'jpg': <__main__.Level1 object at 0x1074863c0>, 'webp': <__main__.Level2 object at 0x107486210>}
```

No need for `__init_subclass__` with what appears to be the same type of data stored.

Reply:
> > Tracking subtypes and their supported format type at the class level adds complexity ...
> 
> @aaraney, can you elaborate further? Assuming for a moment that we are tracking those details, some amount complexity has to go somewhere. Is it drastically more by putting it here than in another class/object (which we'd then have to worry explicitly about making available everywhere it needed to be)?

Right, if we want to keep a mapping from format type to format detectors and vice versa, lets assume doing that has some complexity that would more or less be the same no matter how you implement it. The difference I want to point out, is you can accomplish the same goal and do so in a manor that is easier to understand and reason with. So, the complexity difference is the _cognitive_ complexity that is added by moving this abstraction to this level rather than moving it out into its own abstraction. That is a little hand wavy. In a more cut and dry sense, I think having a singleton that we manually register the format to format detector mapping is easier to understand, maintain, and use in the long run. What you have now _is_ a singleton, however consequently each subtype and therefore instance all reference what are effectively singleton dictionaries. 
 
> > ... and overly couples a type to its behavior.
> 
> I don't understand. Are you talking about somehow coupling the superclass to specific subclass behavior?

The relationship between a subclass of a `ItemDataDomainDetector` and a data format is prescribed to have at most, one format type. This coupling at the type definition level structurally couples what that type should be used for and makes it difficult if not impossible to refactor and maintain backwards compatibility in the future. Tracking of the relationships between a format and a format detector by the super class only further complicates refactoring. While these concepts are all related and will be used together in tandem to provide detection capabilities, IMO loosening the coupling in code between the concept of a detector and a detector registry or detector factory will make our lives easier in the future. 


> > Like @christophertubbs alluded to, due to the nature of python imports, if a subtype is not parsed during runtime, it will not be tracked by the super type.
> 
> This is true, but so is the reverse. We can take advantage of this to loosen dependency constraints and dynamically utilize new subclasses.

This is what I want to avoid. Although, what you suggested sounds nice now, I would much rather have to manually register formats to format detectors on something like a singleton format detector registry. While it is not the most elegant solution, I think it is the most clear and easy to understand in passing. This problem reminds me a lot of an http mux registry. So you can imagine my example maps to more or less how django handles registering a handler for some path. We could also pursue using a decorator pattern to register and make the connection between a format and a format detector. The point being, IMO registering a format detector to a format should be clear and obvious.

> > I recommend that if a format detector factory is necessary, we create that abstraction where it is needed, preferably not in the core.dataset module.
> 
> ...
> 
> As for a factory, I think I started that way but ran into a wall: I don't think there is a feasible way to decide whether a subclass is the subclass we want to create until after we've created it and had it detect a domain (or fail to do so).
>
> The primary thing we have to represent the general form of some data is the DataFormat, so a detector factory would need to use format to pick what subclass of instance to produce. The format would have to be provided to the factory by whatever code is using the factory. Thinking a bit higher level for a moment, how will that entity know the data format? Well, it might be provided before then, but I am pretty sure that implies the user pass it via a CLI arg, and the entire point here is that we don't want to have to rely on that.

I should have said registry instead of factory. Sorry if I added confusion by being imprecise! Maybe i'm missing something, but I don't see any reason why the functionality this class provides through registration and lookup couldn't be provided through a standalone format detection registry? That of course would not include automating registration. But, all the functionality written here should be feasible, yeah?

With that in mind, I don't see why couldn't the implementation of the `UniversalItemDomainDetector` couldn't just take an instance of a format detection registry and use it for what it does now.

Reply:
> I would much rather have to manually register formats to format detectors on something like a singleton format detector registry

Having written clever solutions to avoid manual registration, at this point I'd recommend it.  Non-manual registration adds a little too much magic (which we already have a lot of). The manual registration pattern is already something that Python already designs against in other places (notably multiprocessing context managers), so, despite having to deal with the pain point of having that code someplace and making sure it's always hit, it already has a reasonable precedent.